[ZIP 307] Update out-of-date light client protocol definition. (#1202)
zcash/zips: zips/zip-0307.rst

Reviewer: This definition seems exclusive of "view-key", and maybe other, (reduced capability?) clients.

Aside:  Are there kinds of "not full participant" clients that are NOT capability reductions of "spend-receive" clients?

I think what's confusing to me is the word "full".

It could be read as having to do with validation, which the light-client protocol explicitly does not support, but a reader might find it confusing that a "not full participant" is implied to be able to "send and receive".

In other words a view-only lightclient is certainly NOT a "full participant in the network of Zcash peers" AND it cannot "send and receive".

Proposed edit:

_A client that is not a full participant in the network of Zcash peers. It does not store or validate a full copy of the block chain._

Proposal:

A more nuanced lexicon for reduced-or-divergent-capability clients.

Reviewer 2: ```suggestion
A client that is not a full participant in the network of Zcash peers. It can receive
and (if it holds the required spending keys) send payments, but does not store or
validate a full copy of the block chain.
```

Terminology
===========

The key words "MUST", "SHOULD", "SHOULD NOT", and "MAY" in this document are to be
interpreted as described in BCP 14 [#BCP14]_ when, and only when, they appear in all
capitals.

> 

Reviewer 2: ```suggestion

The character § is used when referring to sections of the Zcash Protocol Specification. [^protocol]

The terms "Mainnet" and "Testnet" are to be interpreted as described in § 3.12 ‘Mainnet and Testnet’.
[^protocol-networks]

```

Abstract
========

This ZIP defines a protocol for Zcash light clients that enables detection of incoming
payments, identification of spent notes, and maintenance of note commitment tree witnesses
for Sapling [#zip-0205]_ and Orchard [#zip-0224]_ shielded transactions, as well as

> transparent balance tracking.

Reviewer 2: ```suggestion
and nullifier sets for Sapling [#zip-0205]_ and Orchard [#zip-0224]_ shielded transactions,
as well as transparent input and output tracing.
```

Motivation
==========

Prior to the introduction of this protocol, a client that wished to send or
receive shielded payments had to be a full node participating in the Zcash
network. This required an amount of available bandwidth, space, and processing

> power that was unsuitable for some classes of user. This light client protocol

Reviewer 2: ```suggestion
power that was unsuitable for some use cases. This light client protocol
```

addresses that need, and is appropriate for low-power, bandwidth-conscious, or

> otherwise limited machines (such as mobile phones).

Reviewer 2: "Most" involves a subjective weighting: do you count use cases by number, importance, number of users? "Some" is sufficient.

> - **Zcash node** that provides chain state and serves as a root of trust for the system.

Reviewer 2: ```suggestion
- **Zcash full node** that provides chain state and serves as a root of trust for the system.
```

> - **Proxy server** that extracts block chain data from a Zcash node to store and serve it

Reviewer: We use "Indexer" widely:

Proposals:

"Indexing Proxy"

OR

"Indexer"

I prefer more explicit.

are specified in the ``compact_formats.proto`` file of the Zcash Light Client Protocol
[#lightwallet-protocol]_.

The protocol has evolved through several major versions:

- **v0.3.x** (2021–2023): Added Orchard shielded protocol support via
  ``CompactOrchardAction`` and note commitment tree state queries
  (``GetTreeState``, ``GetLatestTreeState``) and subtree root retrieval
  (``GetSubtreeRoots``) to support efficient witness construction.

- **v0.4.x** (2025): Added transparent transaction data (``CompactTxIn``, ``TxOut``) to
  compact blocks, enabling transparent outputs to be discovered without revealing
  the client's addresses to the light wallet server. The update also added
  ``PoolType``-based filtering for selective retrieval by shielded protocol.

The following sections describe the rationale for the compact format design.

Output Compression
------------------

In the Zcash protocol, each shielded output contains an AEAD-encrypted *note
plaintext* [#protocol-notept]_. For both Sapling and Orchard, this encrypted
ciphertext (``encCiphertext``) is 580 bytes long and includes the memo field

> and an authentication tag.

Reviewer 2: ```suggestion
plaintext* [#protocol-notept]_. For both Sapling and Orchard, this ciphertext
(``encCiphertext``) is 580 bytes long and includes the memo field and an
authentication tag.
```

  ``PoolType``-based filtering for selective retrieval by shielded protocol.

The following sections describe the rationale for the compact format design.

Output Compression
------------------

In the Zcash protocol, each shielded output contains an AEAD-encrypted *note
plaintext* [#protocol-notept]_. For both Sapling and Orchard, this ciphertext
(``encCiphertext``) is 580 bytes long and includes the memo field and an
authentication tag.

A recipient detects their transactions by trial-decrypting this ciphertext. On
a full node that has the entire block chain, the primary cost is computational.
For light clients however, there is an additional bandwidth cost: every
ciphertext on the block chain must be received from the server (or network
node) the light client is connected to.

However, payment detection does not require the full ciphertext. The first 52
bytes of ``encCiphertext`` contain the contents and opening of the note
commitment, which is sufficient to detect, verify, and spend the note. The
compact representation therefore includes only these 52 bytes along with the
note commitment and ephemeral public key.

> For **Sapling**, a ``CompactSaplingOutput`` contains the 32-byte note commitment

Reviewer 2: ```suggestion
For **Sapling**, a ``CompactSaplingOutput`` contains the 32-byte extracted note commitment
```

($\mathsf{cmu}$), the 32-byte ephemeral public key, and the first 52 bytes of
the ciphertext, for a total of 116 bytes per output. A full Sapling Output
description [#protocol-outputencodingandconsensus]_ is 948 bytes in v4
transactions (756 bytes in v5), so the compact representation provides an
85–88% reduction in bandwidth.

For **Orchard**, a ``CompactOrchardAction`` contains the 32-byte nullifier, the

> 32-byte note commitment ($\mathsf{cmx}$), the 32-byte ephemeral public key, and

Reviewer 2: ```suggestion
32-byte extracted note commitment ($\mathsf{cmx}$), the 32-byte ephemeral public key, and
```

Local processing
----------------

Given a ``CompactBlock`` at block height $\mathsf{height}$ received from a proxy server, a
light client processes it in the following ways:

Scanning for relevant transactions
``````````````````````````````````
For every ``CompactSaplingOutput`` and ``CompactOrchardAction`` in the ``CompactBlock``,
the light client attempts trial decryption against its set of incoming viewing keys. The
following subsections describe the compact trial decryption procedures for each protocol.

Sapling compact trial decryption
''''''''''''''''''''''''''''''''

For each ``CompactSaplingOutput`` $(\mathtt{cmu}, \mathtt{ephemeralKey}, \mathsf{ciphertext})$,
the light client trial-decrypts with each Sapling incoming viewing key $\mathsf{ivk}$. The
procedure is a slight deviation from the standard decryption process

> [#protocol-decryptivk]_ (all constants and algorithms are as defined there):

Reviewer 2: Fixed. This ZIP now only defines the diff from 4.20.2 in the protocol specification, which means that ZIP 2005 does not need to update it (memo bundles or the AssetBase addition would need to update it for the compact output ciphertext length).

- if $\mathsf{LEBS2OSP}_{256}(\mathsf{cm}_u') \neq \mathtt{cmu}$, return $\bot$
- return $\mathbf{np}$.

Orchard compact trial decryption
'''''''''''''''''''''''''''''''''

For each ``CompactOrchardAction`` $(\mathtt{nullifier}, \mathtt{cmx}, \mathtt{ephemeralKey},
\mathsf{ciphertext})$, the light client trial-decrypts with each Orchard incoming viewing
key $\mathsf{ivk}$. The procedure is analogous to Sapling but uses Pallas-based key
agreement [#protocol-concreteorchardkeyagreement]_ and note commitment

> [#protocol-decryptivk]_ (all constants and algorithms are as defined there):

Reviewer 2: Fixed.

[#tor]_ or the Nym mixnet [#nym]_.

In addition, even when using a privacy-preserving network connection for these
operations, a client SHOULD NOT make transaction data requests to the proxy
server in a fashion that links the client's interest in that transaction to
other operations that the client may be performing in the clear. For example,
if a client uses an unprotected connection for compact block download, that
client SHOULD NOT retrieve full transaction data immediately upon transaction
discovery, but SHOULD defer it until a later time when the download will 
not be identifiably related to the download of the compact block range in 
which the transaction was discovered.

Importing a pre-existing key
`````````````````````````````
The light client's scanning start height (often called the "wallet birthday")
determines how far back the light client must scan to detect all relevant
transactions. When a light client is initialized with a newly generated key, it
can use the current chain tip as its wallet birthday, since no prior
transactions can be relevant.

When initialized with a pre-existing key —- whether a seed, a full viewing key,

> or an incoming viewing key —- the light client MUST scan from the key's

Reviewer 2: ```suggestion
When initialized with a pre-existing key —whether a seed, a full viewing key,
or an incoming viewing key— the light client MUST scan from the key's
```

.. [#BCP14] `Information on BCP 14 — "RFC 2119: Key words for use in RFCs to Indicate Requirement Levels" and "RFC 8174: Ambiguity of Uppercase vs Lowercase in RFC 2119 Key Words" <https://www.rfc-editor.org/info/bcp14>`_

> .. [#protocol-merkletree] `Zcash Protocol Specification, Version 2025.6.1 [NU6.1]. Section 3.8: Note Commitment Trees <protocol/protocol.pdf#merkletree>`_

Reviewer 2: ```suggestion
.. [#protocol] `Zcash Protocol Specification, Version 2025.6.1 [NU6.1] or later <protocol/protocol.pdf>`_
.. [#protocol-merkletree] `Zcash Protocol Specification, Version 2025.6.1 [NU6.1]. Section 3.8: Note Commitment Trees <protocol/protocol.pdf#merkletree>`_
.. [#protocol-networks] `Zcash Protocol Specification, Version 2025.6.1 [NU6.1]. Section 3.12: Mainnet and Testnet <protocol/protocol.pdf#networks>`_
```
